torch/_inductor/codegen/cpp_gemm_template.py

```diff
@@ -247,6 +265,12 @@ def __init__(
         self.m, self.n, self.k = m, n, k
         self.padded_n = get_padded_n(n, self.register_blocking.block_n)
         self.is_dynamic_M = has_free_symbols((m,))
+        self.is_int4_woq_gemm = is_int4_woq_gemm
+        self.qBlockSize = (
+            (input_nodes[1].get_numel() * 16) / input_nodes[2].get_numel()
+            if self.is_int4_woq_gemm
+            else 0
+        )
```

Comment:
I'm a bit concerned about adding more and more complexity to the gemm template while we can probably refactor the gemm template class a bit to ease extensions. Is it possible to create a subclass to handle woq int4 specific stuff?

Reply:
The only special-case for GEMM template is int4 due to pre-packed weights. If I create a separate class for it, even then the changes in `GEMM_TEMPLATE` would be required.

Moreover, it'd lead to duplication of a lot of code.

I could still proceed if you'd like

Reply:
Is it possible to make the base GEMM template extensible? Not sure why there would be a lot of dup code. But I'm fine to refactor things later though.

Reply:
Thanks, I'll refactor it later because currently, `add_choices` is a static method, so I'll move some of its nested methods outside & make them static methods as well. This way, a derived class could reuse some of those methods without duplicating code.